Better icon theme handling

linkmauve opened the pull request:
We previously were going through a list of hardcoded paths, but this was causing issues such as no handling for different themes (not supported in this commit), or different preferences in directories when the scale factor changes for instance.

linkmauve committed:
Better icon theme handling

We previously were going through a list of hardcoded paths, but this was
causing issues such as no handling for different themes (not supported
in this commit), or different preferences in directories when the scale
factor changes for instance.

chrisduerr commented on src/xdg.rs:
```diff
@@ -221,42 +208,125 @@ impl Icon {
 
+/// Guessed types of images.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+enum Bucket {
```
```suggestion
enum ImageType {
```

chrisduerr commented on src/xdg.rs:
```diff
@@ -221,42 +208,125 @@ impl Icon {
-                if let Some((name, _)) = name {
-                    let _ = icons.insert(name.to_owned(), file.path());
+                // Each directory in an icon theme has the same layout, so let’s hardcode it.
+                let bucket = if file_name == "scalable" {
```
```suggestion
                let image_type = if file_name == "scalable" {
```

chrisduerr commented on src/xdg.rs:
```diff
@@ -221,42 +208,125 @@ impl Icon {
+                    Bucket::Scalable
+                } else if file_name == "symbolic" {
+                    Bucket::Symbolic
+                } else if let [width, height] = file_name.split('x').collect::<Vec<_>>()[..] {
```
split_once will simplify this code.

chrisduerr commented on src/xdg.rs:
```diff
@@ -221,42 +208,125 @@ impl Icon {
+
+            for dir_entry in fs::read_dir(icons_dir).unwrap() {
+                let dir_entry = dir_entry.unwrap();
+                let file_name = dir_entry.file_name();
+                let file_name = match file_name.to_str() {
+                    Some(file_name) => file_name,
+                    None => continue,
+                };
```
Not sure this should be called "file_name". It's technically a directory we're talking about here.

chrisduerr commented on src/xdg.rs:
```diff
@@ -221,42 +208,125 @@ impl Icon {
+
+                let mut dir_path = dir_entry.path();
+                dir_path.push("apps");
+                let read_dir = match fs::read_dir(dir_path) {
+                    Ok(read_dir) => read_dir,
+                    Err(_) => continue,
+                };
```
That variable name doesn't really make sense, should probably be `dir`?

chrisduerr commented on src/xdg.rs:
```diff
@@ -221,42 +208,125 @@ impl Icon {
+                    Ok(read_dir) => read_dir,
+                    Err(_) => continue,
+                };
+                for file in read_dir {
+                    let file = file.unwrap();
```
Should probably flatten here rather than unwrap. Currently this can crash.

linkmauve replied:
I went with a `try_continue!()` macro to remove all of the `.unwrap()` calls.

chrisduerr commented on src/xdg.rs:
```diff
@@ -221,42 +208,125 @@ impl Icon {
+                        Some(file_name) => file_name,
+                        None => continue,
+                    };
+                    let name = file_name.rsplit_once('.');
+                    if let Some((name, _)) = name {
```
Should be inlined probably.

chrisduerr commented on src/xdg.rs:
```diff
@@ -221,42 +208,125 @@ impl Icon {
+                            .or_default()
+                            .insert(Bucket::Scalable, file.path());
+                    },
+                    Some((name, "png")) => {
+                        // We don’t have any information about the size of the icon here.
+                        icons
+                            .entry(name.to_owned())
+                            .or_default()
+                            .insert(Bucket::Unknown, file.path());
+                    },
```
I don't think `Unknown` is a good name here. Just call it bitmap or something. We know the image type we just don't know its size.

chrisduerr committed:
Remove macros

chrisduerr committed:
Improve code style

chrisduerr committed:
Improve code structure

chrisduerr commented on src/xdg.rs:
```diff
@@ -332,7 +389,30 @@ impl IconLoader {
+            // Get all available icons matching this icon name.
+            let matching_icons = self.icons.get(icon).ok_or(Error::NotFound)?;
+
+            // Sort icons by best match for the desired size.
+            let mut matching_icons: Vec<_> = matching_icons.iter().collect();
+            matching_icons.sort_by_key(|(&icon_type, _)| match icon_type {
+                ImageType::SizedBitmap(icon_size) => match icon_size.cmp(&size) {
+                    // If the size of the PNG is identical to ours, it’s the best option.
+                    Ordering::Equal => 0,
+                    // Otherwise order by closest size bigger than ours.
+                    Ordering::Greater => icon_size - size,
+                    // And only prefer smaller sizes after unknown sizes.
+                    Ordering::Less => 8192 + size - icon_size,
+                },
+                // Scalable is obviously the second best option to get crisp icons.
+                ImageType::Scalable => 1,
+                // Here we take the bet that an unknown file will be better than a known smaller.
+                ImageType::Bitmap => 8192,
+                // And finally, symbolic is the same icon but without colours, usually too dark.
+                ImageType::Symbolic => 16384,
+            });
```
I applied all the changes that were pretty non-controversial to speed things up a bit, but with this specific code I'd like to get your feedback first.

First I'm curious what the purpose of sorting is. The "ideal" icon can always be determined in a single iteration, correct? Even to ensure the `vec` is sorted, you'd have to iterate over it anyway so it wouldn't help you. And since it's created from a `HashMap` it's rather unlikely that this is already sorted?

So if we just iterate over them all once, we can save ourselves the sorting and it should avoid having to re-allocate to go from `HashMap` to `Vec`.

Besides that, I also think that this would probably be a good separate method on `self`. It's so long due to the (useful) documentation, that having it in a separate method with a succinct name would likely make reading of this `load` function much easier.

linkmauve replied:
Thanks for the improvements, they are all good!

I have just pushed a new commit which goes a bit further into what I had in mind for the sort, basically doing it once at initial load time and then only sorting again when the size changes.  `IconLoader` isn’t aware of the size until `load()` so we currently have to keep doing the load every time, not only on size changes, but that’s for a future refactor.

chrisduerr commented:
I removed all the macros since they just tend to make code much less readable. Going from `to_str` to `into_string` also allowed avoiding intermediate variables for all the last path segment accesses.

linkmauve committed:
Sort icons by best matching size on first load

This way further sorts (during resize) are very cheap, in the best case
just a single iteration.

chrisduerr commented on src/xdg.rs:
```diff
@@ -323,13 +411,20 @@ impl IconLoader {
 
     /// Load image file as RGBA buffer.
-    fn load(&self, icon: &str, size: u32) -> Result<Icon, Error> {
+    fn load(&mut self, icon: &str, size: u32) -> Result<Icon, Error> {
```
This function is currently only called by `set_scale_factor`, which reloads all icons when the scale factor has changed.

Since the `set_scale_factor` doesn't do anything unless the factor actually changed, this function should never be called with a `size` that is already cached.

So isn't the sorting rather useless? The only help it would be is if an icon was loaded multiple times by different desktop files, but even then why would you sort the list of icons and load it again rather than just loading the icon from the cache?

linkmauve replied:
How else would you do it?

chrisduerr replied:
If the sorting isn't helpful, it's more efficient to just iterate over it with a simple reduce.

chrisduerr committed:
Remove sorting

chrisduerr commented:
I've decided to go ahead and remove the sorting, since the benefit of it was questionable at best and even a pre-sorted list wouldn't drastically improve performance (I doubt users have thousands of icons for one app).

The new code should hopefully be simpler to understand and still behave the same way. The only major change I've made is to prefer known-size smaller bitmaps over bitmaps of unknown sizes. While that was largely for implementation convenience, I think it's a fair change considering the pixmaps are kind of a second class citizen anyway.

I'll leave this open until tomorrow if you have any suggestions or concerns @linkmauve.
